.github/workflows/ci.yml

```diff
@@ -65,5 +65,56 @@ jobs:
       - name: Type-check
         run: pnpm run typecheck
 
+      - name: Check formatting (root)
+        run: npm run format
+        
       - name: Build
         run: pnpm run build
+        
+  update-nix-hash:
+    runs-on: ubuntu-latest
+    if: github.event_name == 'push' && github.ref == 'refs/heads/main'
+    permissions:
+      contents: write  # Needed for push access
+      pull-requests: write  # Needed for PR creation
```

Comment:
Just flagging this for awareness, this job introduces write access from CI for the first time (excluding the CLA Assistant, which already operates with elevated perms via `pull_request_target`).

It’s scoped well (only runs on push to main) so it won’t run on PRs or forks. That said, it does slightly shift the trust boundary, merging into main will now trigger a job that can push branches and open PRs back to the repo.

No action needed here, but it might be worth making sure folks reviewing automation changes are aware of that shift. If main isn’t already protected, enabling “require PRs before merging” and disallowing force-pushes/deletion would be good hardening steps for this kind of setup.

Reply:
since this job is the only one in ci.yml with write permissions, it might be worth considering moving it to a separate workflow file (e.g. `update-nix-hash.yml`). That could make it easier to see at a glance which workflows mutate the repo and which are strictly read-only CI.

Also worth noting, the script `update-nix-hash.sh` will be run with elevated permissions and therefore also needs special attention during review to ensure it is not going to smuggle a secret, accidentally commit something, and that it’s not vulnerable to input injection (e.g. using PR metadata like branch names in shell commands without quoting).